libs/ml/include/ml/layers/fully_connected.hpp

```diff
@@ -75,43 +75,65 @@ class FullyConnected : public SubGraph<T>
+    , init_mode_(init_mode)
   {
     // get correct name for the layer
     std::string name = GetName();
 
     // start to set up the structure
-    std::string input =
-        this->template AddNode<fetch::ml::ops::PlaceHolder<TensorType>>(name + "_Input", {});
-
+    input_ = this->template AddNode<fetch::ml::ops::PlaceHolder<TensorType>>(name + "_Input", {});
     // for non time distributed layer, flatten the input
-    std::string flat_input = input;
+    flat_input_ = input_;
     if (!time_distributed_)
     {
-      flat_input =
-          this->template AddNode<fetch::ml::ops::Flatten<TensorType>>(name + "_Flatten", {input});
+      flat_input_ =
+          this->template AddNode<fetch::ml::ops::Flatten<TensorType>>(name + "_Flatten", {input_});
     }
 
-    std::string weights =
-        this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Weights", {});
-    std::string weights_matmul = this->template AddNode<fetch::ml::ops::MatrixMultiply<TensorType>>(
-        name + "_MatrixMultiply", {weights, flat_input});
-    std::string bias =
-        this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Bias", {});
-    std::string output = this->template AddNode<fetch::ml::ops::Add<TensorType>>(
-        name + "_Add", {weights_matmul, bias});
+    weights_ = this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Weights", {});
+
+    weights_matmul_ = this->template AddNode<fetch::ml::ops::MatrixMultiply<TensorType>>(
+        name + "_MatrixMultiply", {weights_, flat_input_});
+
+    bias_ = this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Bias", {});
+
+    add_ = this->template AddNode<fetch::ml::ops::Add<TensorType>>(name + "_Add",
+                                                                   {weights_matmul_, bias_});
 
-    output = fetch::ml::details::AddActivationNode<T>(activation_type, this, name + "_Activation",
-                                                      output);
+    output_ =
+        fetch::ml::details::AddActivationNode<T>(activation_type, this, name + "_Activation", add_);
 
-    this->AddInputNode(input);
-    this->SetOutputNode(output);
-    this->SetRegularisation(fetch::ml::details::CreateRegulariser<T>(regulariser),
-                            regularisation_rate);
+    this->AddInputNode(input_);
+    this->SetOutputNode(output_);
+  }
+
+  void CompleteInitialisation() override
+  {
+    assert(!this->expected_slice_input_shapes_.empty());
+    assert(!this->slice_output_shape_.empty());
+    this->nodes_.at(input_)->SetExpectedSliceInputShapes(this->expected_slice_input_shapes_);
+    this->nodes_.at(input_)->SetSliceOutputShape(this->expected_slice_input_shapes_.front());
+    FETCH_LOG_INFO(DESCRIPTOR, "-- Compiling sub-graph ... --");
+
+    // When expected input shape is known, it is possible to compute flattened input
+    // shape of this fully connected layer
+    this->nodes_.at(flat_input_)
```

Comment:
flat_input_ is a confusing naming choice given how it's now being used

Reply:
input_ and flat_input_ could be the same if the real input Tensor is 2-dimensional (e.g. data is plain 1 dim vector) and could be 2 different layers if this Dense is time_distributed (I don't know what does it mean, but the code says so :) 
I think we could name this node "flattened_input_" or alike.